Added TypeScript type checker + Fixed type errors.

sainthkh opened the pull request:
- Closes #5772

### What has been changed?

#### 1. Added type checker  done by `tsc`.

![Screenshot from 2019-11-25 11-16-58](https://user-images.githubusercontent.com/8130013/69507554-4c5de180-0f76-11ea-8d54-5cfe921a4bae.png)

![Peek 2020-01-16 10-38](https://user-images.githubusercontent.com/8130013/72490610-0a0c9e80-385b-11ea-9261-e829f787e963.gif)

`npm run type-check` command iterates folders in `packages` and check if it has `tsconfig.json` file and runs `tsc` command to check if there is a type error in files. It ignores `cli` folder because `dtslint` checks the types. 

It doesn't completely remove `dtslint` for now because it cannot check the parameter or return type of a variable. 

This command is added to CI.

**Command Options**
* -p, --project: To list projects separated by commas (like `reporter,runner`).
* --skip-lib-check: --skipLibCheck of tsc.
* --ignore-progress: Ignores progress for CI. [Without this, CI result looks like this.](https://circleci.com/gh/cypress-io/cypress/196854?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

#### 2. Other code changes.

* Removed copying `@types` from `cli`. It causes duplicate types and creates a lot of errors. Fixed a new feature introduced in [TypeScript 2.9, `import('npm-module')`](https://stackoverflow.com/questions/39040108/import-class-in-definition-file-d-ts).  
* Moved `@types` dependencies of `cli` from `devDependencies` to `dependencies`. Because they are used in `types/index.d.ts`.
* Updated packages to fix type errors. 
  * @types/chai: 4.1.7 -> 4.2.5 => We're using chai@4.2, but types weren't updated. 
  * typescript: 3.5.3 -> 3.7.2 => To use `window.NODE` definition.
  * Removed `typeRoots` in reporter, runner `tsconfig.json` files. 

### Additional details

- Why was this change necessary?

Currently, CI doesn't check type errors. 

- What is affected by this change?

When there is a type error, CI will fail.

- Any implementation details to explain?

It uses `tsc --noEmit` to check errors. 

### How has the user experience changed?

N/A

### PR Tasks

- [NA] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [NA] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [NA] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

sainthkh commented:
Note: electron-5 fail is a flaky test.

jennifer-shehane commented:
@sainthkh Some failures in the type checks

```
> node scripts/type_check "--ignore-progress"


desktop-gui failed
node_modules/@cypress/schema-tools/src/api.ts:449:9 - error TS2769: No overload matches this call.
  Overload 1 of 6, '(fn: (x: BindOptions) => CustomFormats | undefined, list: readonly BindOptions[]): (CustomFormats | undefined)[]', gave the following error.
    Argument of type '<T>(obj: Record<"formats", T>) => T' is not assignable to parameter of type '(x: BindOptions) => CustomFormats | undefined'.
      Types of parameters 'obj' and 'x' are incompatible.
        Type 'BindOptions' is not assignable to type 'Record<"formats", CustomFormats | undefined>'.
          Property 'formats' is optional in type 'BindOptions' but required in type 'Record<"formats", CustomFormats | undefined>'.
  Overload 2 of 6, '(fn: (x: number | BindOptions | (() => string) | { (...items: ConcatArray<BindOptions>[]): BindOptions[]; (...items: (BindOptions | ConcatArray<BindOptions>)[]): BindOptions[]; } | ... 17 more ... | (() => IterableIterator<...>)) => any, list: BindOptions[]): readonly any[]', gave the following error.
    Argument of type '<T>(obj: Record<"formats", T>) => T' is not assignable to parameter of type '(x: number | BindOptions | (() => string) | { (...items: ConcatArray<BindOptions>[]): BindOptions[]; (...items: (BindOptions | ConcatArray<BindOptions>)[]): BindOptions[]; } | ((searchElement: BindOptions, fromIndex?: number | undefined) => number) | ((searchElement: BindOptions, fromIndex?: number | undefined) => number) | ((start?: number | undefined, end?: number | undefined) => BindOptions[]) | ((searchElement: BindOptions, fromIndex?: number | undefined) => boolean) | (() => string) | (<U>(callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => U, thisArg?: any) => U[]) | { <S extends BindOptions>(callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => value is S, thisArg?: any): S[]; (callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => unknown, thisArg?: any): BindOptions[]; } | ((callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => unknown, thisArg?: any) => boolean) | ((separator?: string | undefined) => string) | ((callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => unknown, thisArg?: any) => boolean) | ((callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => void, thisArg?: any) => void) | { (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions): BindOptions; (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions, initialValue: BindOptions): BindOptions; <U>(callbackfn: (previousValue: U, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => U, initialValue: U): U; } | { (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions): BindOptions; (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions, initialValue: BindOptions): BindOptions; <U>(callbackfn: (previousValue: U, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => U, initialValue: U): U; } | { <S extends BindOptions>(predicate: (this: void, value: BindOptions, index: number, obj: BindOptions[]) => value is S, thisArg?: any): S | undefined; (predicate: (value: BindOptions, index: number, obj: BindOptions[]) => unknown, thisArg?: any): BindOptions | undefined; } | ((predicate: (value: BindOptions, index: number, obj: BindOptions[]) => unknown, thisArg?: any) => number) | (() => IterableIterator<[number, BindOptions]>) | (() => IterableIterator<number>) | (() => IterableIterator<BindOptions>)) => any'.
      Types of parameters 'obj' and 'x' are incompatible.
        Type 'number | BindOptions | (() => string) | { (...items: ConcatArray<BindOptions>[]): BindOptions[]; (...items: (BindOptions | ConcatArray<BindOptions>)[]): BindOptions[]; } | ((searchElement: BindOptions, fromIndex?: number | undefined) => number) | ((searchElement: BindOptions, fromIndex?: number | undefined) => number) | ((start?: number | undefined, end?: number | undefined) => BindOptions[]) | ((searchElement: BindOptions, fromIndex?: number | undefined) => boolean) | (() => string) | (<U>(callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => U, thisArg?: any) => U[]) | { <S extends BindOptions>(callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => value is S, thisArg?: any): S[]; (callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => unknown, thisArg?: any): BindOptions[]; } | ((callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => unknown, thisArg?: any) => boolean) | ((separator?: string | undefined) => string) | ((callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => unknown, thisArg?: any) => boolean) | ((callbackfn: (value: BindOptions, index: number, array: BindOptions[]) => void, thisArg?: any) => void) | { (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions): BindOptions; (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions, initialValue: BindOptions): BindOptions; <U>(callbackfn: (previousValue: U, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => U, initialValue: U): U; } | { (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions): BindOptions; (callbackfn: (previousValue: BindOptions, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => BindOptions, initialValue: BindOptions): BindOptions; <U>(callbackfn: (previousValue: U, currentValue: BindOptions, currentIndex: number, array: BindOptions[]) => U, initialValue: U): U; } | { <S extends BindOptions>(predicate: (this: void, value: BindOptions, index: number, obj: BindOptions[]) => value is S, thisArg?: any): S | undefined; (predicate: (value: BindOptions, index: number, obj: BindOptions[]) => unknown, thisArg?: any): BindOptions | undefined; } | ((predicate: (value: BindOptions, index: number, obj: BindOptions[]) => unknown, thisArg?: any) => number) | (() => IterableIterator<[number, BindOptions]>) | (() => IterableIterator<number>) | (() => IterableIterator<BindOptions>)' is not assignable to type 'Record<"formats", CustomFormats | undefined>'.
          Type 'number' is not assignable to type 'Record<"formats", CustomFormats | undefined>'.
  Overload 3 of 6, '(fn: (x: BindOptions) => CustomFormats | undefined, obj: Functor<BindOptions>): Functor<CustomFormats | undefined>', gave the following error.
    Argument of type '<T>(obj: Record<"formats", T>) => T' is not assignable to parameter of type '(x: BindOptions) => CustomFormats | undefined'.
      Types of parameters 'obj' and 'x' are incompatible.
        Type 'BindOptions' is not assignable to type 'Record<"formats", CustomFormats | undefined>'.

449     map(prop('formats'), options),
            ~~~~~~~~~~~~~~~


node_modules/@cypress/schema-tools/src/trim.ts:25:13 - error TS2322: Type 'PlainObject' is not assignable to type 'never'.

25             trimmedObj[prop] = trimBySchema(schema[prop], object[prop])
               ~~~~~~~~~~~~~~~~

node_modules/@cypress/schema-tools/src/trim.ts:27:13 - error TS2322: Type 'PlainObject[]' is not assignable to type 'never'.

27             trimmedObj[prop] = map(trimBySchema(schema[prop]), object[prop])
               ~~~~~~~~~~~~~~~~


Found 3 errors.
```

sainthkh commented:
They're from schema-tools. I wrote PR for that. Please check PR comment above.

sainthkh commented:
1. Fixed conflicts. 
2. Tests failed by flaky build.

jennifer-shehane commented:
@sainthkh There's some failures from our dtslint 

```
Error: Errors in typescript@2.9 for external dependencies:
../node_modules/@types/chai/index.d.ts(122,9): error TS8020: JSDoc types can only be used inside documentation comments.
../node_modules/@types/chai/index.d.ts(123,9): error TS8020: JSDoc types can only be used inside documentation comments.
../node_modules/@types/chai/index.d.ts(127,16): error TS2370: A rest parameter must be of an array type.

    at /root/cypress/cli/node_modules/dtslint/bin/index.js:190:19
    at Generator.next (<anonymous>)
    at fulfilled (/root/cypress/cli/node_modules/dtslint/bin/index.js:5:58)
```

sainthkh commented:
Fixed dtslint problem. It happened because `typescript@2.9` cannot understand optional elements in tuple types. [Check here](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/38473)

Fixed this by updating the minimum typescript version to 3.0.

However, Linux lint will fail because `@cypress-io/schema-tools` isn't updated. I fixed the type error in cypress-io/schema-tools#298. But it's not updated to npm.

To get this PR done:

- [ ] Release new version of schema-tools to npm.
- [ ] Release new version of json-schemas to npm. 
- [ ] Update version of `json-schemas` at `packages/desktop-gui`.

Then, the type failures in desktop-gui will be fixed.

jennifer-shehane commented:
@sainthkh I've pinged @tgriesser about the issue with schemat-tools not releasing.

jennifer-shehane commented:
@sainthkh semantic-release won't release if the commit message was just a `chore`, so you needed to commit with `fix` in the commit message. I made an empty commit with the `fix` keyword, gonna see if that releases. https://github.com/cypress-io/schema-tools/commit/c2d3bf7c50d92f950ec80317b2b6a3e0035111e4

sainthkh commented:
schema-tools is now released. Now we need to release json-schema that uses it. 

Then, we can update desktop gui.

jennifer-shehane commented:
The PR to schema-tools is failing here https://github.com/cypress-io/json-schemas/pull/464

sainthkh commented:
I don't have access to json-schema, unfortunately.

jennifer-shehane commented:
5.33.1 of json-schemas has been released with the schema-tools update.

sainthkh commented:
1. Fixed type errors.
2. Failed because of a flaky test.

sainthkh commented:
Type error in desktop-gui:

```
desktop-gui failed
../../node_modules/@cypress/json-schemas/dist/index.d.ts:20:33 - error TS2694: Namespace '"/home/sainthkh/src/cypress/node_modules/@types/ramda/index"' has no exported member 'CurriedFunction2'.

20     getExample: import("ramda").CurriedFunction2<string, string, PlainObject | undefined>;
                                   ~~~~~~~~~~~~~~~~

../../node_modules/@cypress/json-schemas/dist/index.d.ts:23:27 - error TS2694: Namespace '"/home/sainthkh/src/cypress/node_modules/@types/ramda/index"' has no exported member 'CurriedFunction3'.

23     trim: import("ramda").CurriedFunction3<string, string, PlainObject, PlainObject>;
                             ~~~~~~~~~~~~~~~~

../../node_modules/@cypress/json-schemas/dist/index.d.ts:24:32 - error TS2694: Namespace '"/home/sainthkh/src/cypress/node_modules/@types/ramda/index"' has no exported member 'CurriedFunction2'.

24     hasSchema: import("ramda").CurriedFunction2<string, string, boolean>;
                                  ~~~~~~~~~~~~~~~~

../../node_modules/@cypress/json-schemas/dist/index.d.ts:25:27 - error TS2694: Namespace '"/home/sainthkh/src/cypress/node_modules/@types/ramda/index"' has no exported member 'CurriedFunction3'.

25     fill: import("ramda").CurriedFunction3<string, string, object, PlainObject>;
                             ~~~~~~~~~~~~~~~~

```

Bkucera commented:
this doesn't output any files does it? i've seen that running `tsc` locally will output `.js` files next to the `.ts`, which messes up running cypress with `ts-node` require hooks

sainthkh commented:
Yes. It doesn't compile ts files. It just checks types.

Bkucera commented:
we should try to get this PR in soon. This is great work @sainthkh !

sainthkh commented:
* Fixed conflicts. 
* chrome test failed because of a flaky test.
* I don't have access to @cypress/json-schemas. Please fix type errors in that project.

sainthkh commented:
* Merged.
* lint error is caused by @cypress-io/json-schemas
* Other failures are flaky.

jennifer-shehane commented:
@sainthkh Sent an invite to give you permission to @cypress/json-schemas

sainthkh commented:
Opened a PR in cypress-io/json-schemas#526.

Things to do:
- [ ] The PR merged
- [ ] json-schemas new version published
- [ ] update it in desktop-gui

tgriesser commented:
@sainthkh changes should be published under 5.34.1

sainthkh commented:
* Now lint-types passes. 
* 2 flaky tests.

CypressJoseph commented on packages/reporter/src/hooks/hook-model.spec.ts:
```diff
@@ -1,4 +1,5 @@
 import sinon from 'sinon'
+
```
nitpick: stray whitespace maybe?

sainthkh replied:
I might have added it to separate external dependencies and internal ones. But I think it doesn't matter to remove it.

sainthkh commented on circle.yml:
```diff
@@ -318,11 +318,10 @@ jobs:
       - run:
           command: yarn lerna exec  --scope cypress "yarn dtslint"
+      - run:
+          command: yarn type-check --ignore-progress
```
@CypressJoseph I added this to CI. Or is there anything more to do?

CypressJoseph commented:
That’s great! :+1: This should be good to go I think.

On Mon, Mar 16, 2020 at 5:52 AM Kukhyeon Heo <notifications@github.com>
wrote:

> *@sainthkh* commented on this pull request.
> ------------------------------
>
> In circle.yml
> <https://github.com/cypress-io/cypress/pull/5780#discussion_r392896069>:
>
> > +      - run:
> +          command: yarn type-check --ignore-progress
>
> @CypressJoseph <https://github.com/CypressJoseph> I added this to CI. Or
> is there anything more to do?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cypress-io/cypress/pull/5780#pullrequestreview-375052993>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOVVUCT7CMWMSWSVKIBFZ2TRHXZHRANCNFSM4JQZIRKQ>
> .
>

jennifer-shehane commented:
@sainthkh We're planning to merge this into `develop` directly after the 4.2.0 release (which is today), so we can try it out in `develop` without potentially blocking our release.
